File: app/src/main/java/com/example/bookapp/viewModels/ViewModelMessages.java
package com.example.bookapp.viewModels;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.bookapp.models.MessageDTO;
import com.example.dataLayer.repositories.MessageRepository;

import java.util.ArrayList;


public class ViewModelMessages extends ViewModel implements MessageRepository.MessageRepositoryCallback {
    //the last messages in the chat
    private MutableLiveData<ArrayList<MessageDTO>> messages = new MutableLiveData<>();
    private MutableLiveData<ArrayList<MessageDTO>> currentlyFetchedNewMessages = new MutableLiveData<>();
    private MutableLiveData<MessageDTO> currentlySentMessage = new MutableLiveData<>();

    public MutableLiveData<ArrayList<MessageDTO>> getMessages() {
        return messages;
    }

    public MutableLiveData<MessageDTO> getCurrentlySentMessage() {
        return currentlySentMessage;
    }

    public MutableLiveData<ArrayList<MessageDTO>> getCurrentlyFetchedNewMessages() {
        return currentlyFetchedNewMessages;
    }

    /*************************************************INTERFACE METHODS ************************************/

    @Override
    public void onOldMessagesFetched(@NonNull ArrayList<MessageDTO> oldMessageDTOS) {
        ArrayList<MessageDTO> newData = new ArrayList<>();
        if (this.messages.getValue() != null) {
            newData.addAll(this.messages.getValue());
        }
        newData.addAll(oldMessageDTOS);

        this.messages.setValue(newData);
    }


    @Override
    public void onNewMessagesReady(@NonNull ArrayList<MessageDTO> newMessageDTOS) {
        //create a new arrayList to trigger the observer
        if (this.messages.getValue() != null) {
            this.messages.getValue().addAll(newMessageDTOS);
        }
        currentlyFetchedNewMessages.setValue(newMessageDTOS);
    }

    @Override
    public void onSendMessageReady(@NonNull MessageDTO messageDTO) {
        currentlySentMessage.setValue(messageDTO);
    }


}

File: app/src/main/java/com/example/bookapp/viewModels/ViewModelMessages.kt
package com.example.bookapp.viewModels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import com.example.bookapp.models.UserMessage
import com.example.dataLayer.repositories.MessageRepository
import kotlinx.coroutines.InternalCoroutinesApi
import kotlinx.coroutines.launch

@InternalCoroutinesApi
class ViewModelMessages(application: Application) : AndroidViewModel(application) {


    private val messageRepository: MessageRepository = MessageRepository(viewModelScope, application = getApplication())

    fun getRecentMessages(currentUserID: String, user2ID: String): LiveData<List<UserMessage>> {
        if (messageRepository.requests[user2ID] == null) {
            viewModelScope.launch {
                messageRepository.fetchRecentMessages(currentUserID, user2ID);
            }
        }
        return messageRepository.getMessages(currentUserID, user2ID)
    }

    fun requestMoreMessages(currentUserID: String, user2ID: String,currentOffset:Int) {
        viewModelScope.launch {
            messageRepository.fetchMoreMessages(currentUserID, user2ID,currentOffset)
        }
    }

}